File: prolog-frontend/test/lists.pl
append([], Ys, Ys).
append([X|Xs], Ys, [X|Zs]) :- append(Xs, Ys, Zs).

member(X, [X|_]).
member(X, [_|Xs]) :- member(X, Xs).

length([], 0).
length([_|T], N) :- length(T, N1), N is N1 + 1.

reverse(Xs, Ys) :- reverse(Xs, [], Ys).
reverse([], Acc, Acc).
reverse([X|Xs], Acc, Ys) :- reverse(Xs, [X|Acc], Ys).

last([X], X).
last([_|T], X) :- last(T, X).

File: erlang-frontend/test/lists.erl
-module(lists).
-export([len/1, map/2, filter/2, reverse/1, append/2]).

len([]) -> 0;
len([_ | T]) -> 1 + len(T).

map(_, []) -> [];
map(F, [H | T]) -> [F(H) | map(F, T)].

filter(_, []) -> [];
filter(F, [H | T]) ->
    case F(H) of
        true -> [H | filter(F, T)];
        false -> filter(F, T)
    end.

reverse(L) -> reverse(L, []).
reverse([], Acc) -> Acc;
reverse([H | T], Acc) -> reverse(T, [H | Acc]).

append([], B) -> B;
append([H | T], B) -> [H | append(T, B)].
